### DISCUSSION TITLE
Semantyczne wersjonowanie w GitFlow

### DISCUSSION THREAD

**Discussion by roofiq** [AzureCommunityPL/group#151] [Pomoc] [answered]
Witam!

Chciałbym się poradzić o implementację semantycznego wersjonowania w środku projektu (chyba to nie ma znaczenia, ale aktualnie jest już działająca wersja aplikacji na produkcji). Pracujemy w metodologii GitFlow.

Problem: zbudować CI pipeline dla solucji, który również zaimplementuje semantyczne wersjonowanie. Wszystko w Azure DevOps.

Do implementacji wersjonowania znalazłem dwa możliwe podejścia:

1. [GitVersion](https://gitversion.net/docs) - jest to narzędzie, w którym konfigurujemy plik `GitVersion.yml` i według dokumentacji, w łatwy sposób możemy go skalibrować pod nasze potrzeby.
Bawiłem się dość długo tym narzędziem i muszę przyznać, że dokumentacja jest dosyć biedna. Również nawet z ilości "Issues" na GitHubie można odebrać wrażenie, że nie jest to dość dojrzałe rozwiązanie i jest pełno niedociągnięć.
Przy testach, nie udało mi się inkrementować wersji "Patcha" przy merge'owaniu HotFixa. Patch zwiększał się tylko o +1 i tylko przy commit'cie "+semver: patch". Następnych inkrementacji nie udało się osiągnąć.
Utworzyłem o tym case na ich GitHubie, gdzie można poczytać więcej - [https://github.com/GitTools/GitVersion/issues/2454](https://github.com/GitTools/GitVersion/issues/2454)

2.  Drugim podejściem byłaby samodzielna implementacja logiki wersjonowania w pipeline'ie. Zapewne musiałby to być jakiś skrypt PowerShellowy, który w zależności od Branch'u i Trigger'u inkrementował dany parametr.

Jak to działa u Was w projektach? Czy używacie jakiegoś 3rd party silnika czy może wpisujecie wartości z "ręki"?

Dzięki!
Rafał

**Comment by kamil-mrzyglod** [comment, marked as answer]
Pamiętam, że widziałem GitVersion i także mi nie podpasował. Generalnie problemem jest brak natywnego narzędzia, które w jasny sposób automatyzuje wersjonowanie. Ja bym rozróżnił problem w zalezności czy:
* wersjonujesz artefakty
* wersjonujesz binarki

Czyli inaczej - czy wystarczy wersjonowanie na poziomie build / release czy musisz grzebać w kodzie aby podmienić wersję paczki / assembly. Ja osobiście widzę jeden problem z poleganiem na nomenklaturze w treści commita - musisz zbudować w zespołach nawyk aby faktycznie oznaczali jak wpływ na wersję ma zmiana. Można to pewnie nieco uprościć wymuszając strukturę z użyciem np. git hook, ale znów - musisz na każdej maszynie tego hooka odpalić.

Jeśli miałbym to zrobić na teraz z użyciem git flow to pewnie bym sobie odpuścił pluginy do ADO(może dlatego, że nie znam dobrej) i napisał kilka linijek kodu, które by decydowały jak podbić wersję w zależności od brancha. Jeśli wykorzystujecie Boards i linki do work itemów to można by było zintegrować to z tagami aby decydować która zmiana jest major / minor / patch.

To wszystko też przy założeniu, że wersjonujesz binarkę. Na wersjonowanie builda / releasa w ten sposób nie znam innej metody jak manualne ustawianie bądź rzeźba z nadpisywanie Build.BuildNumber gdzie nawet nie mam pewności, że to się da :)

**Comment by piotrgwiazda** [comment]
Hej
W większości przypadków używam dwuczłonowwej wersji. Wersja "semantyczna"
np. 1.2.0, 3.4.1 itd ustawiana jakoś ręcznie. Albo jest w jakimś pliku w
źródłach, albo np. brana jest z nazwy brancha (release/1.2.0).
Do tego doklejam numer builda generowany przez AzDO. Razem:
1.2.0-2020116.2

W trunk-based development nie ma brancha z numerem. Czasem całkiem ignoruję
wersję semantyczną i używam tylko numer builda.

Natknąłem się też na rózne pluginy co bazują na tagach i szukają zmian od
ostatniego taga. Niektóre potrafią też z monorepo zbudować tylko zmienione
kawałki

https://github.com/qoomon/maven-git-versioning-extension

https://github.com/vackosar/gitflow-incremental-builder

https://danielflower.github.io/multi-module-maven-release-plugin/usage.html

Te są akurat okołojavowe, ale pewnie do .NET też coś się znajdzie.
Ewentualnie może Bazel ma jakieś wypaśniejsze pluginy w razie
skomplikowanego procesu.




niedz., 15 lis 2020, 21:36 użytkownik Kamil Mrzygłód <
notifications@github.com> napisał:

**Comment by roofiq** [comment]
Dzięki Panowie!
Chodzi o wersjonowanie artefaktów. Jednak też porzuciłem GitVersion i załatwiłem temat skryptem :)
Ogólnie może trochę podciągnę dyskusję o używanie Artifacts Feeds. Dwa pytania:
1. Restorowanie paczek - czy robicie to przez task do Restoru czy może przez  [ściągnięcie paczek z Build'a](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/download-build-artifacts?view=azure-devops) żeby zaoszczędzić na czasie? Myślałem, że jeśli będziemy je cache'ować w Feedzie to przyspieszy proces ale niestety tak to nie działa.
Brakuje mi Tasku, który pozwoliłby ściągnąć wszystkie paczki z danego Feeda - aktualnie trzeba wybrać po jednej paczce ([DownloadPackage](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/utility/download-package?view=azure-devops)) - chyba, że znowu coś napisać customowego :)

2. Jesteście zwolennikami Push'owania do Feeda tylko stablinych wersji czy może korzystacie z promowania? Zastanawiam się, którą drogą pójść, poniżej wklejam notkę z [docsów](https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?view=azure-devops&tabs=yaml), która opisuje obie
```
When you're producing a package in a build, you can't know whether it will be the version that you aim to release to your users or just a step along the way toward that release. Although none of the following solutions are ideal, you can use one of these depending on your preference:

- After you validate a package and decide to release it, produce another package without the prerelease label and publish it. The drawback of this approach is that you have to validate the new package again, and it might uncover new issues.

- Publish only packages that you want to release. In this case, you won't use a prerelease label for every build. Instead, you'll reuse the same package version for all packages. Because you do not publish packages from every build, you do not cause a conflict.

```

Dzięki za odpowiedzi!

**Reply by kamil-mrzyglod** [reply]
Co do cache'owania paczek - można tutaj nieco pokombinować. Teoretycznie masz do użycia cache task - https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops. Alternatywą jest użycie self-hosted agenta, który nie umiera po każdym uruchomieniu + możesz zaimplementować tam swoją logikę. W zależności od skali może to mieć mniej lub więcej sensu.

Na drugie pytanie nie znam dobrej odpowiedzi, tj. innej niż "to zależy". Jeśli w Twoim procesie uwzględniasz wersje preview, to pewnie jest sens promowania. To komplikuje proces i nie zawsze jest łatwe / możliwe do wykonania(przykład: wersja jest przyspawana w ramach binarek, jej zmiana wymusza ponowne zbudowanie aplikacji co trochę mija się z celem - o tym też pisze @piotrgwiazda poniżej). Ja często korzystałem jednak z pushowania tylko stabilnych wersji i promowania ich np. pomiędzy środowiskami bez zmiany wersji.

**Comment by piotrgwiazda** [comment]
Ogólne zasady, których się trzymam:
- Deployment na produkcję musi być tej samej paczki, binarnie tego samego
pliku, który testowałem. Żadnych rebuildów "z tego samego kodu" ale innego
taga/brancha. Ściągną się nowsze wersje zależności albo coś i to już nie
jest ta sama wersja.
- Każdy build ma unikalny numer wersji i nie ma możliwości nadpisania
paczki nową sztuką. Nowy build, nowa wersja, nowa paczka.
- Promuję wersję/paczkę/builda a nie brancha.
- Stare paczki trzeba jakimś skryptem kasować, a te co poszły na prod
archiwizować jakoś


wt., 17 lis 2020, 23:22 użytkownik roofiq <notifications@github.com>
napisał:
